flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java

```diff
@@ -0,0 +1,703 @@
+            Savepoint lastCheckpoint,
+            Context<FlinkBlueGreenDeployment> josdkContext,
+            boolean isFirstDeployment)
+            throws JsonProcessingException {
+
+        deploy(bgDeployment, nextDeploymentType, lastCheckpoint, josdkContext, isFirstDeployment);
+
+        setAbortTimestamp(bgDeployment, deploymentStatus);
+
+        return patchStatusUpdateControl(
+                        bgDeployment, deploymentStatus, nextState, JobStatus.RECONCILING)
+                .rescheduleAfter(getReconciliationReschedInterval(bgDeployment));
+    }
+
+    private boolean isDeploymentReady(
+            FlinkDeployment deployment,
+            Context<FlinkBlueGreenDeployment> josdkContext,
+            FlinkBlueGreenDeploymentStatus deploymentStatus) {
+        if (ResourceLifecycleState.STABLE == deployment.getStatus().getLifecycleState()
+                && JobStatus.RUNNING == deployment.getStatus().getJobStatus().getState()) {
+            // TODO: checking for running pods seems to be redundant, check if this can be removed
+            int notRunningPods =
+                    (int)
+                            getDeploymentPods(josdkContext, deployment)
+                                    .filter(
+                                            p ->
+                                                    !p.get()
+                                                            .getStatus()
+                                                            .getPhase()
+                                                            .equalsIgnoreCase("RUNNING"))
+                                    .count();
+
+            if (notRunningPods > 0) {
+                LOG.warn("Waiting for " + notRunningPods + " Pods to transition to RUNNING status");
+            }
+
+            return notRunningPods == 0;
+        }
+
+        logPotentialWarnings(
+                deployment, josdkContext, deploymentStatus.getLastReconciledTimestamp());
+        return false;
+    }
+
+    private static Stream<PodResource> getDeploymentPods(
+            Context<FlinkBlueGreenDeployment> josdkContext, FlinkDeployment deployment) {
+        var namespace = deployment.getMetadata().getNamespace();
+        var deploymentName = deployment.getMetadata().getName();
+
+        return josdkContext
+                .getClient()
+                .pods()
+                .inNamespace(namespace)
+                .withLabel("app", deploymentName)
+                .resources();
+    }
+
+    private boolean hasSpecChanged(
+            FlinkBlueGreenDeploymentSpec newSpec, FlinkBlueGreenDeploymentStatus deploymentStatus) {
+
+        String lastReconciledSpec = deploymentStatus.getLastReconciledSpec();
+        String newSpecSerialized = SpecUtils.serializeObject(newSpec, "spec");
+
+        // TODO: in FLIP-504 check here the TransitionMode has not been changed
+
+        return !lastReconciledSpec.equals(newSpecSerialized);
```

Comment:
We may want to use the DiffBuilder/ existing utilities to get the type of the spec change here and make sure we don't trigger deployments on non-upgrade changes. 

Things like operator config changes etc should not go through the Blue/Green flow I believe but simply applied on the currently active deployment.